core/sql/generator/GenPreCode.cpp

```diff
@@ -11274,6 +11274,485 @@ short HbaseAccess::extractHbaseFilterPreds(Generator * generator,
+	for (ValueId vid = preds.init();
+       (preds.next(vid));
+       preds.advance(vid))
+    {
+      ItemExpr * ie = vid.getItemExpr();
+
+      // if it is AND operation, recurse through left and right children
+      if (ie->getOperatorType() == ITM_AND)
+        {
+          ValueIdSet leftPreds;
+          ValueIdSet rightPreds;
+          leftPreds += ie->child(0)->castToItemExpr()->getValueId();
+          rightPreds += ie->child(1)->castToItemExpr()->getValueId();
+          if (isFirstAndLayer){
+            NABoolean leftOK  = extractHbaseFilterPredsV2(generator, leftPreds, newExePreds,TRUE, TRUE);
+            NABoolean rightOK = extractHbaseFilterPredsV2(generator, rightPreds, newExePreds,TRUE, TRUE);
+            if (leftOK && rightOK){
+            	if (!checkOnly){
+					extractHbaseFilterPredsV2(generator, leftPreds, newExePreds,FALSE, TRUE);//generate tree
+					extractHbaseFilterPredsV2(generator, rightPreds, newExePreds,FALSE, TRUE);//generate tree
+					opList_.insert("AND"); // insert an AND node since both side are OK to push down
+					if (addedNode>0)opList_.insert("AND"); // if it is not the first node addd to the push down, AND it with the rest
+					addedNode++; // we just pushed it down, so increase the node count pushed down.
+            	}
+            	if (preds.entries()==1)
+            		return TRUE;
+            } else if (leftOK){  // if only left is OK to push down
+            	if(!checkOnly){
+					extractHbaseFilterPredsV2(generator, leftPreds, newExePreds,FALSE, TRUE);//generate left tree
+					newExePreds.insert(rightPreds); //make sure we add the right child to predicates that needs executor evalvaluation
+					if (addedNode>0)opList_.insert("AND"); // if it is not the first node addd to the push down, AND it with the rest
+					addedNode++; // we pushed down left side so mark it
+            	}
+            	if (preds.entries()==1)
+            		return TRUE;
+            } else if (rightOK){// if only right is OK to push down
+            	if(!checkOnly){
+					extractHbaseFilterPredsV2(generator, rightPreds, newExePreds,FALSE, TRUE);//generate right tree
+					newExePreds.insert(leftPreds);//make sure we add the left child to predicates that needs executor evalvaluation
+					if (addedNode>0)opList_.insert("AND"); // if it is not the first node addd to the push down, AND it with the rest
+					addedNode++;// we pushed down right side so mark it
+            	}
+            	if (preds.entries()==1)
+            		return TRUE;
+            } else{
+            	if(!checkOnly){
+            		newExePreds.insert(vid);// we pushed down nothing, make sure the whole node is evaluated by Executor
+            	}
+            	if (preds.entries()==1)
+            		return FALSE;
+
+            }
+          }
+          else{//if not first AND layer, both left and right must be pushable to get anything pushed
+              if(extractHbaseFilterPredsV2(generator, leftPreds, newExePreds, TRUE, FALSE)&&
```

Comment:
I had the same fear but was thinking that compiler must have a reasonable limit to not push too big or list and turn it into a hash check instead of cycling to many values by for example turning things into hash join? My assumption is not correct? If not the solution will not be to turn that function into no recursive, but block the push down all together, as it is going to be counter productive. Having a huge list of interpreted OR filter is definitely not a good idea and should not be pushed down. Instead, we would want to do a implement a new hbase filter to do hash style filter, or dicotomic check, all in one single filter.